[DK Unholy] Font+Coral Interaction when used together

nyterage opened the pull request:
Small changes to the interaction between font and razor coral when utilized together. 
T24 profile with CLF and Unholy frenzy gets about a 100 sim dps increase. 
T24 profile with CLF and Gargoyle gets about a 150 sim dps increase. 
Vision of Perfection with unholy frenzy gets around a 250 sim dps increase.
verified other talent combinations with both essences remain unchanged. 

Also added a minor change to CLF when utilizing Army of the Damned talent to hold CLF for army, very small dps gain in 5 minute sims. 

full results can be seen here
https://www.raidbots.com/simbot/report/vkRD33PvDZ3UaSN2Aes9LQ

nyterage committed:
Font+Coral Interaction when used together

Melekus commented:
Feel free to split the on-use behavior into multiple lines, these statements are quite a hassle to read and I think you're missing a few cases.

All your essence.name.equipped should probably be changed to essence.name.major, or they'll return true even if only the minor is equipped.

From what I could understand, Azshara's font of power on-use covers:
- CLF + Ashvane's
- end of fight + CLF (might also need an explicit CLF major check)
- no CLF or perfection major (used on cooldown, intended?)
- perfection without gargoyle
- perfection with gargoyle (used on cooldown, intended?)

A missing case is where CLF major is equipped but Ashvane's razor coral is not.

Ashvane's razor coral usage has the same issues, the case covered are the following:
- CLF and apoc pets are both active
- CLF is not equipped, army of the damned and unholy frenzy aren't talented, apoc cd < gcd
- CLF is equipped, fight is ending before CLF comes up, apoc cd < gcd
- CLF is equipped, fight is ending before apoc comes up
- perfection is equipped, fight is ending before apoc comes up
- Unholy frenzy buff is up, CLF and perfection aren't equipped.

This seems to be missing several cases
- perfection is equipped/major, and UF or army of the damned is talented (there's only one use for the end of the fight, after the last full activation of apocalypse, intended?)
- neither perfection nor CLF are major, UF is not talented but army of the damned is talented.

nyterage commented:
yeah, they are a bit of a mess to read, sorry about that, ill clean it up by separating it out a bit. 

my mistake not putting .major and instead using .active for the essences, will fix that up as well

as for the Azshara's font of power lines, yeah you got that just right, it is also intended for font to be used on cooldown in the cases not specified (same behavior as the current APL basically)

For razor coral, its kind of just a mess for specific scenarios, but you got that correct as well. 

those missing cases ill add in, sorry for missing those, the APL for these trinkets is messy so i must have forgotten to include them.

nyterage committed:
Easier to read & adding some forgotten elements

nyterage commented:
so i believe that fixed the issues you had previously. 

1. made it much easier to read, separated each essence/non essence specific line. 

Azsharas:

1. changed the end of fight modifier to work better with any essence
2. Left out a CLF with no coral, as the damage gain from holding font rather than using on cooldown is marginal.
3. changed some code up to interact nicely with minor essences as well.

Razor Coral:

1. Changes made to play nicely with minors.
2. Tested several modifiers with Vision and Army of the Damned, all resulted in a DPS loss, reverted to utilizing it every apoc available (if Azsharas is utilized first, it will not utilize coral, this is intended). 
3. Same situation with Army of the Damned without CLF or Vision, utilizing coral every apoc results in the best dps as far as i could manage, leaving as is. 


As for "perfection is equipped/major, and UF or army of the damned is talented (there's only one use for the end of the fight, after the last full activation of apocalypse, intended?)" 
there was a line in there for apoc cd if vision is enabled, must have been messed up somehow, i apologize for how much of a cluster it was. 

updated sim results:
https://www.raidbots.com/simbot/report/qw3Zq7ySjBxSwHDSbTu53f

if there are any other issues you notice, please let me know! 

EDIT: should also note, the ones with coral being utilized only on the final apoc of the fight is intended, its odd, but results in the highest damage. here's an alternative Sim without that interaction. Vision buids seem to suffer the most, loosing around 50dps by not utilizing coral end of fight only. 

https://www.raidbots.com/simbot/report/p6w62TEdeE7N3iEsSBVyz8

the version of the APL without coral end of fight apoc only may be better for all sim lengths, and a 50dps loss vs potential on 5 minute sims isnt a major difference. current APL utilizes this interaction as well for Vision Gargoyle and Vision AoTD it seems.

Melekus commented on engine/class_modules/sc_death_knight.cpp:
```diff
@@ -7824,9 +7824,13 @@ void death_knight_t::default_apl_unholy()
-  def -> add_action( "use_item,name=azsharas_font_of_power,if=(cooldown.apocalypse.remains<15&essence.condensed_lifeforce.enabled&equipped.ashvanes_razor_coral)|(target.1.time_to_die<cooldown.condensed_lifeforce.remains+30)|(!essence.condensed_lifeforce.enabled&!essence.vision_of_perfection.enabled)|(essence.vision_of_perfection.enabled&cooldown.apocalypse.remains<14&!talent.summon_gargoyle.enabl ...
+  def -> add_action( "use_item,name=azsharas_font_of_power,if=(cooldown.apocalypse.remains<=14&essence.condensed_lifeforce.major&equipped.ashvanes_razor_coral)|(target.1.time_to_die<cooldown.apocalypse.remains+34)" );
+  def -> add_action( "use_item,name=azsharas_font_of_power,if=(!essence.condensed_lifeforce.major&!essence.vision_of_perfection.major)|(talent.army_of_the_damned.enabled&essence.vision_of_perfection.enabled)" );
+  def -> add_action( "use_item,name=azsharas_font_of_power,if=(essence.vision_of_perfection.major&talent.summon_gargoyle.enabled)|(essence.vision_of_perfection.major&cooldown.apocalypse.remains<14&!talent.summon_gargoyle.enabled)" );
```
You have 6 use cases here, but that can be reduced down to 3.

No CLF + No Perfection, perfection + aotd, perfection + gargoyle. The trinket is used on cooldown in all three cases.
CLF major, perfection major + unholy frenzy. The trinket is used with apocalypse cooldown < 14 in both cases.

``` 
use_item,name=azsharas_font_of_power,if=(essence.vision_of_perfection.major&!talent.unholy_frenzy.enabled)|(!essence.condensed_lifeforce.major&!essence.vision_of_perfection.major)
use_item,name=azsharas_font_of_power,if=cooldown.apocalypse.remains<14&(essence.condensed_lifeforce.major|essence.vision_of_perfection.major&talent.unholy_frenzy.enabled)
use_item,name=azsharas_font_of_power,if=target.1.time_to_die<cooldown.apocalypse.remains+34
```

Simplifying it that way would make it more readable in my opinion.

Chewsdaydk commented:
use_item,name=ashvanes_razor_coral,if=(cooldown.apocalypse.remains<gcd&!essence.condensed_lifeforce.major&!essence.vision_of_perfection.major)|(buff.unholy_frenzy.up&!essence.condensed_lifeforce.major&!essence.vision_of_perfection.major)

I believe the Unholy Frenzy half of this line is redundant - when Ashvane's is equipped Unholy Frenzy can be thought of as a 90 second cooldown and should be handled as such - Thus the Apocalypse condition will always trigger before the Frenzy condition. Please let me know if i've missed something.

Melekus commented on engine/class_modules/sc_death_knight.cpp:
```diff
@@ -7824,9 +7824,13 @@ void death_knight_t::default_apl_unholy()
-  def -> add_action( "use_item,name=ashvanes_razor_coral,if=(pet.guardian_of_azeroth.active&pet.apoc_ghoul.active)|(cooldown.apocalypse.remains<gcd&!essence.condensed_lifeforce.enabled&!talent.army_of_the_damned.enabled&!talent.unholy_frenzy.enabled)|(cooldown.apocalypse.remains<gcd&target.1.time_to_die<cooldown.condensed_lifeforce.remains+20)|(target.1.time_to_die<cooldown.apocalypse.remains+20& ...
+  def -> add_action( "use_item,name=ashvanes_razor_coral,if=(pet.guardian_of_azeroth.active&pet.apoc_ghoul.active)|(cooldown.apocalypse.remains<gcd&target.1.time_to_die<cooldown.condensed_lifeforce.remains+20&essence.condensed_lifeforce.major)|(target.1.time_to_die<cooldown.apocalypse.remains+20&essence.condensed_lifeforce.major)" );
+  def -> add_action( "use_item,name=ashvanes_razor_coral,if=(cooldown.apocalypse.remains<gcd&!essence.condensed_lifeforce.major&!essence.vision_of_perfection.major)|(buff.unholy_frenzy.up&!essence.condensed_lifeforce.major&!essence.vision_of_perfection.major)" );
+  def -> add_action( "use_item,name=ashvanes_razor_coral,if=(cooldown.apocalypse.remains<gcd&essence.vision_of_perfection.major)|(target.1.time_to_die<cooldown.apocalypse.remains+20&essence.vision_of_perfection.major)" );
```
I reformulated in a similar manner as the previous chunk

```
use_item,name=ashvanes_razor_coral,if=pet.guardian_of_azeroth.active&pet.apoc_ghoul.active
use_item,name=ashvanes_razor_coral,if=cooldown.apocalypse.ready&(essence.condensed_lifeforce.major&target.1.time_to_die<cooldown.condensed_lifeforce.remains+20|!essence.condensed_lifeforce.major)
use_item,name=ashvanes_razor_coral,if=target.1.time_to_die<cooldown.apocalypse.remains+20&(essence.condensed_lifeforce.major|essence.vision_of_perfection.major)
```
Is it necessary to keep the essence check on the last target.1.time_to_die check?
I removed the Unholy Frenzy check as it seemed unnecessary following Chews' comment.
I also replaced the cooldown.apocalypse.remains<gcd with a simple cooldown.apocalypse.ready

nyterage committed:
Updating as per Melekus and Chewie's feedback

nyterage commented:
ah, you guys are both entirely correct, guess i was too caught up in the use cases i didn't see how simplified it could be. 

updated to reflect your suggested changes.

nyterage committed:
remove major conditions on time to die

nyterage committed:
fix a copy/paste fail
